src/cplscheme/BaseCouplingScheme.hpp

```diff
@@ -545,6 +476,88 @@ class BaseCouplingScheme : public CouplingScheme {
 
   /// Writes out coupling convergence within all time windows.
   std::shared_ptr<io::TXTTableWriter> _convergenceWriter;
+
+  /// First participant name.
+  std::string _firstParticipant = "unknown";
+
+  /// Second participant name.
+  std::string _secondParticipant = "unknown";
+
+  /// Local participant name.
+  std::string _localParticipant = "unknown";
+
+  /// Smallest number, taking validDigits into account: eps = std::pow(10.0, -1 * validDigits)
+  const double _eps;
+
+  /**
+   * @brief All convergence measures of coupling iterations.
+   *
+   * Before initialization, only dataID and measure variables are filled. Then,
+   * the data is fetched from send and receive data assigned to the cpl scheme.
+   */
+  std::vector<ConvergenceMeasure> _convergenceMeasures;
+
+  /// Sets whether the solver evaluates the fine or the coarse model representation
+  bool _isCoarseModelOptimizationActive = false;
+
+  /// Functions needed for initialize()
+
+  /// implements functionality needed by initialize if __couplingMode == Implicit
+  virtual void initializeImplicit() = 0;
+
+  /// implements functionality for initialize in base class.
+  virtual void initializeImplementation() = 0;
+
+  /// Functions needed for initializeData()
+
+  /// implements functionality for initializeData in base class.
+  virtual void initializeDataImpl() = 0;
+
+  /// Functions needed for advance()
+
+  /// implements functionality for advance in base class.
+  virtual void explicitAdvance() = 0;
+
+  /// implements functionality for advance in base class.
+  virtual std::pair<bool, bool> implicitAdvance() = 0;
+
+  /// If any required actions are open, an error message is issued.
+  void checkCompletenessRequiredActions();
+
+  /// Returns true if end time of time window is reached. Does not check for convergence
+  bool subcyclingIsCompleted();
```

Comment:
What about "isWindowCompleted" ?

Reply:
The term subcycling is not ideal, since it does not cover what we have in mind for multirate timestepping (#133). However, `isWindowCompleted` sounds very similar to [`bool isTimeWindowComplete()`](https://github.com/precice/precice/blob/9660959ecb73c91082e594f950f3c807aee3747a/src/precice/SolverInterface.hpp#L243), which has different meaning:

* `bool subcyclingIsCompleted()` (this function): Are we at the end of the time window? Does not tell us anything about whether coupling has converged.
* `bool isTimeWindowComplete()`: Are we done with computations for this window? I.e. copuling converged **and** we have reached the end of the window.

I think we should come up with something that is more concrete than `Complete`.

Reply:
List of spontaneous ideas:

* `atEndOfTimeWindow`
* `isEndOfTimeWindowReached`
* `isTimeWindowEndReached`
* `reachedTimeWindowEnd`
* `timeIsAtEndOfWindow`
* `timeReachedEndOfWindow`

If we use something with the term `TimeWindowEnd` it might be a good idea to have an additional member (or function), `_timeWindowEnd` (or `timeWindowEnd()`). This makes naming more clear. 

However, it might also require us to revisit some implementations of functions where we currently compare `_computedTimeWindowPart` and `_timeWindowSize` (i.e. `getThisTimeWindowRemainder()`). We could instead compare `_time` and `timeWindowEnd`. This puts everything from a framework that is relative to the `_timeWindowStart` to a framework of absolute numbers.

Reply:
9d5ea56ed2e99ee83c7ec9d50b58f9ab02bc7980 changes the name to `reachedEndOfTimeWindow`. Having a new member like `_timeWindowEnd` would require us to add many more members, functions etc. I am not sure whether this is really worth it just for the sake of naming.

I expect that we will have to revisit how we are tracking time when dealing with #133, anyway. And we do not know yet, what exactly we need. Therefore, there is a high risk that we implement something that we will change very soon again. I would suggest to leave the whole revision of time tracking to the future and don't touch it yet.

@uekerman if you agree with the current solution, feel free to mark this conversation as resolved.